### PR TITLE
Add documentation to document event sync subscriber

### LINKED ISSUE #36416: Add documentation to document event sync subscriber
**Description:**
$Ttile

**Describe your task(s)**

**Related Issues (optional):**
#35057 

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by gayalkuruppu**
Fixes #36416 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [x] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+This document contains information about the implementation of a notification system inside the language server using the 
+PubSub architecture. Goal is to allow communication between language server and extensions synchronously.
+
+Following diagram shows how this has been implemented. 
+
+![alt text](../images/language-server/001-DocumentEventPubSub-class-diagram.png)
+
+- Event Subscribers and Event Publishers get loaded using JavaSPI and get subscribed to publishers when a document 
+opens or changes.
+- Event Publishers and Subscribers have an Event kind and a subscriber can only subscribe to a single publisher at once.
```
Comment (mohanvive) [inline docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md:14]: EventKind related info is not clear for me.

File: docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+If we need a subscriber to subscribe for two publishers, then we have to create two different subscribers and subscribe
```
Comment (mohanvive) [inline docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md:15]: Don't show the link between publisher and subscriber. Publisher should link to Event Kind. Also, Subscriber should link to Event Kind.

Comment (mohanvive) [reply docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md:15]: For example, 

If we need to consume events from multiple EventKinds, then we have to create two different subscribers and subscribe to those EventKinds individually.

File: docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+separately.
+- When an Event Publisher publishes, the `onEvent()` method is getting triggered for each subscriber that has been
+subscribed to the publisher.
```
Comment (mohanvive) [inline docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md:18]: Ideal to have, something like,

- When an Event Publisher publishes events to an EventKind, the `onEvent()` method is getting triggered for each subscriber that has been subscribed to the specific EventKind

**Review by mohanvive — CHANGES_REQUESTED** [review]

File: docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md (hunk 1 above)
Comment (IMS94) [inline docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md:14]: We can say `EventKind acts as the topic. A subscriber can be notified on only 1 EventKind`.
A subscriber is subscribing to the event kind, not to the publisher

File: docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+## Overview
+
+This document contains information about the implementation of a notification system inside the language server using the 
+PubSub architecture. Goal is to allow communication between language server and extensions synchronously.
+
+Following diagram shows how this has been implemented. 
+
+![alt text](../images/language-server/001-DocumentEventPubSub-class-diagram.png)
+
+- Event Subscribers and Event Publishers get loaded using JavaSPI and get subscribed to publishers when a document
```
Comment (IMS94) [inline docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md:12]:
```suggestion
- Event Subscribers and Event Publishers get loaded using JavaSPI and subscribed to an EventKind. A subscriber is notified when a publisher publishes an event (document update, open, close, etc) under that EventKind 
```

File: docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+If we need a subscriber to subscribe for two publishers, then we have to create two different subscribers and subscribe
+separately.
+- When an Event Publisher publishes, the `onEvent()` method is getting triggered for each subscriber that has been
+subscribed to the publisher.
+- We can override `publish()` and `onEvent()` methods to control the event synchronization.
+  
+
+## How to Use
+
+- **Event Publisher** is an internal interface and cannot be accessed publicly by extension developers. The server
```
Comment (IMS94) [inline docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md:24]:
```suggestion
- **Event Publisher** is an internal interface which is not exported from the module. The server 
```

File: docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+- When an Event Publisher publishes, the `onEvent()` method is getting triggered for each subscriber that has been
+subscribed to the publisher.
+- We can override `publish()` and `onEvent()` methods to control the event synchronization.
+  
+
+## How to Use
+
+- **Event Publisher** is an internal interface and cannot be accessed publicly by extension developers. The server 
+decides which events to be delegated to a particular publisher. In the `publish()` method a debouncing mechanism has 
+implemented for the **Project Update Event Publisher**.
```
Comment (IMS94) [inline docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md:26]: Classes we should use `ClassName` formatting. Not bold font. Same for the `EventPublisher` and `EventSubscriber` interfaces.

File: docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+    @Override
+    public EventKind eventKind() {
+        return EventKind.PROJECT_UPDATE;
+    }
+    
+    @Override
+    public void onEvent(ExtendedLanguageClient client, DocumentServiceContext context,
+                        LanguageServerContext languageServerContext) {
+        checkAndRegisterCommands(context);
```
Comment (IMS94) [inline docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md:51]:
```suggestion
        // Your implementation here
```
No need to copy the same logic here. Just put a link to an implementation class (like CommandRegisteringSubscriber`.

File: docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+    
+    public static synchronized void checkAndRegisterCommands(DocumentServiceContext context) {
```
Comment (IMS94) [inline docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md:59]: Remove this method

File: docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md (hunk 7 above)
Comment (IMS94) [reply docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md:51]: Also mention that the `DocumentServiceContext` and the `fileUri` can be used to determine `scheme` specific events. Currently LS supports `file` and `expr` schemes.

File: docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    @Override
+    public void publish(ExtendedLanguageClient client, LanguageServerContext serverContext,
+                        DocumentServiceContext context) {
+        if (latestScheduled != null && !latestScheduled.isDone()) {
```
Comment (IMS94) [inline docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md:103]: No need to have an implementation here. Just say `// your implementation here` and below this provide a link to an existing implementation.

**Review by IMS94 — CHANGES_REQUESTED** [review]
Added few comments. Please move the `images` directory into the `language-server` directory

File: docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md (hunk 4 above)
Comment (gayalkuruppu) [reply docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md:12]: For now, we are not publishing events on `didClose` and we do not have to publish events on `didClose` right? @IMS94

Comment (gayalkuruppu) [reply docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md:12]: Can you be more specific on the `image` folder move? currently the image is in the `docs/images/language-server/001-DocumentEventPubSub-class-diagram.png` path

Comment (IMS94) [reply docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md:12]: Let's add LS related images in `docs/language-server/images` folder. We have to create that.

Comment (gayalkuruppu) [reply docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md:12]: what about the `didClose` @IMS94

Comment (IMS94) [reply docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md]: Though we don't support `didClose` as an event, it's a possible extension in the future. Just wanted the reader to get the idea of event kinds and subscribing.

File: docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        // Your implementation here
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+}
+```
+[`CommandRegisterSubscriber`](../../language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/eventsync/subscribers/CommandRegisterSubscriber.java)
```
Comment (IMS94) [inline docs/language-server/DocumentEventSyncPublisherSubscriberDesign.md:62]:
```suggestion
See [`CommandRegisterSubscriber`](../../language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/eventsync/subscribers/CommandRegisterSubscriber.java) for an example
```

File: docs/language-server/DocumentEventSyncPublisherSubscriberDeveloperGuide.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+                        DocumentServiceContext context,
+                        LanguageServerContext languageServerContext) {
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+}
+```
```
Comment (IMS94) [inline docs/language-server/DocumentEventSyncPublisherSubscriberDeveloperGuide.md:43]: Add a link to an example here as well:
```
See [`CommandRegisterSubscriber`](../../language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/eventsync/subscribers/CommandRegisterSubscriber.java) for an example
```

**Review by IMS94 — CHANGES_REQUESTED** [review]

**Comment by gayalkuruppu** [comment]
@IMS94 did not include a mermaid diagram since I did not find a way to group classes/interfaces for ls-core and ls-commons inside a rectangle (except for `loop`, `alt`, `opt` and `subgraph`)

File: docs/language-server/DocumentEventSyncPublisherSubscriberDeveloperGuide.md [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+`EventSubscriber`s. 
+- You will get the following log message in VSCode if you have created and subscribed to the specific publisher 
+correctly as in the example. 
+```
+Trigger service subscriber subscribed to Project update event publisher
+```
+
+#### 2. Unit Testing
+
+- Unit tests are written inside `EventSyncPubSubTest`.
```
Comment (IMS94) [inline docs/language-server/DocumentEventSyncPublisherSubscriberDeveloperGuide.md:73]: `EventSyncPubSubTest` is just an example right? Does it cover how to write unit tests for a new subscriber?

**Comment by IMS94** [comment]
> @IMS94 did not include a mermaid diagram since I did not find a way to group classes/interfaces for ls-core and ls-commons inside a rectangle (except for `loop`, `alt`, `opt` and `subgraph`)

No problem.

File: docs/language-server/DocumentEventSyncPublisherSubscriberDeveloperGuide.md (hunk 12 above)
Comment (gayalkuruppu) [reply docs/language-server/DocumentEventSyncPublisherSubscriberDeveloperGuide.md:73]: It does not. `EventSyncPubSubTest` just contains current unit test written to test `EventSyncTestSubscriber` being subscribed to `PROJECT_UPDATE` Event.

Comment (IMS94) [reply docs/language-server/DocumentEventSyncPublisherSubscriberDeveloperGuide.md:73]: Then we have to mention it like that. Like
```
See EventSyncPubSubTest for an example of writing unit tests for subscribers
```

**Review by IMS94 — APPROVED** [review]

**Review by mohanvive — APPROVED** [review]

**Pull request merged**